cvmfs/swissknife_graft.cc

```diff
@@ -0,0 +1,249 @@
+    return 1;
+  }
+
+  if (input_file != "-") {
+    bool input_file_is_dir = (0 == stat(input_file.c_str(), &sbuf)) &&
+                            S_ISDIR(sbuf.st_mode);
+    if (input_file_is_dir) {
+      if (!output_file_is_dir && output_file.size()) {
+        std::cerr << "Input (" << input_file << ") is a directory but output (" << output_file << ") is not" << std::endl;
+        return 1;
+      }
+      if (m_verbose) {
+        std::cerr << "Recursing into directory " << input_file << std::endl;
+      }
+      return Recurse(input_file, output_file);
+    } else {
+      return Publish(input_file, output_file, output_file_is_dir, false);
+    }
+  }
+  return Publish(input_file, output_file, output_file_is_dir, true);
+}
+
+
+int swissknife::CommandGraft::Publish(const std::string &input_file, const std::string &output_file, bool output_file_is_dir, bool input_file_is_stdin) {
+  if (output_file.size() && m_verbose) {
+    std::cout << "Grafting " << input_file << " to " << output_file << std::endl;
+  } else if (!output_file.size()) {
+    std::cout << "Grafting " << input_file << std::endl;
+  }
+  int fd;
+  if (input_file_is_stdin) {
+    fd = 0;
+  } else {
+    fd = open(input_file.c_str(), O_RDONLY, 0755);
+    if (fd < 0) {
+      std::stringstream ss; ss << "Unable to open input file (" << input_file << ")";
+      perror(ss.str().c_str());
+      return 1;
+    }
+  }
+
+  shash::Any hash(shash::kSha1);
+  off_t processed_size;
+  int retval = ChecksumFd(fd, hash, processed_size);
+  if (!input_file_is_stdin) {close(fd);}
+  if (retval) {
+    std::stringstream ss; ss << "Unable to checksum input file (" << input_file << ")";
+    perror(ss.str().c_str());
+    return 1;
+  }
+
+  // Build the .cvmfsgraft-$filename
+  if (output_file.size()) {
+    std::string dirname, fname;
+    std::string graft_fname;
+    if (output_file_is_dir) {
+      splitpath(input_file, dirname, fname);
+      graft_fname =  output_file + "/.cvmfsgraft-" + fname;
+    } else {
+      splitpath(output_file, dirname, fname);
+      graft_fname = dirname + "/.cvmfsgraft-" + fname;
+    }
+    fd = open(graft_fname.c_str(), O_CREAT|O_TRUNC|O_WRONLY, 0755);
+    if (fd < 0) {
+      if ((errno == EISDIR) && !input_file_is_stdin) {
+        splitpath(input_file, dirname, fname);
+        graft_fname = output_file + "/.cvmfsgraft-" + fname;
+        fd = open(graft_fname.c_str(), O_CREAT|O_TRUNC|O_WRONLY, 0755);
+      }
+      if (fd < 0) {
+        std::stringstream ss; ss << "Unable to open graft file (" << graft_fname << ")";
+        perror(ss.str().c_str());
+        return 1;
+      }
+    }
+  } else {
+    fd = 1;
+  }
+  std::stringstream ss;
+  ss << "size=" << processed_size << std::endl;
+  ss << "checksum=" << hash.ToString(true) << std::endl;
+  std::string graft_contents = ss.str();
+  size_t nbytes = graft_contents.size();
+  const char *buf = graft_contents.c_str();
+  while (nbytes) {
```

Comment:
You mentioned this point several times. I think we largely ignored `EINTR` in the code so far.  This might have been a mistake.  Is `EINTR` something that actually happens in practice?

In any case, I'd prefer fixing `CopyMem2File` and other spots.  We could, for instance, add in util.cc `SafeRead` and `SafeWrite` wrappers for `read` and `write`.  Just not sure how to unit test this.